PC/python_uwp.cpp

```diff
@@ -46,170 +49,158 @@ set_user_base()
-        if (wcscpy_s(buffer, len, installPath.c_str())) {
-            Py_FatalError("failed to copy to buffer");
-            return NULL;
-        }
-        if (wcscat_s(buffer, len, L"\\")) {
-            Py_FatalError("failed to concatenate backslash");
-            return NULL;
-        }
-    } else {
-        if (wcscpy_s(buffer, len, argv0)) {
-            Py_FatalError("failed to copy argv[0]");
-            return NULL;
+static int
+set_process_name(PyConfig *config)
+{
+    std::wstring executable, home;
+
+    const auto family = get_package_family();
+    
+    if (!family.empty()) {
+        PWSTR localAppData;
+        if (SUCCEEDED(SHGetKnownFolderPath(FOLDERID_LocalAppData, 0,
+                                           NULL, &localAppData))) {
+            executable = std::wstring(localAppData)
+                         + L"\\Microsoft\\WindowsApps\\"
+                         + family
+                         + L"\\"
+                         + PROGNAME;
+
+            CoTaskMemFree(localAppData);
         }
+    }
 
-        wchar_t *name = wcsrchr(buffer, L'\\');
-        if (name) {
-            name[1] = L'\0';
+    home.resize(MAX_PATH);
+    while (true) {
+        DWORD len = GetModuleFileNameW(
+            NULL, home.data(), (DWORD)home.size());
+        if (len == 0) {
+            home.clear();
+            break;
+        } else if (len == home.size() &&
+                   GetLastError() == ERROR_INSUFFICIENT_BUFFER) {
+            home.resize(len * 2);
         } else {
-            buffer[0] = L'\0';
+            home.resize(len);
+            size_t bslash = home.find_last_of(L"/\\");
+            if (bslash != std::wstring::npos) {
+                home.erase(bslash);
+            }
+            break;
         }
     }
 
-    if (wcscat_s(buffer, len, PROGNAME)) {
-        Py_FatalError("failed to concatenate program name");
-        return NULL;
+    if (executable.empty() && !home.empty()) {
+        executable = home + L"\\" + PROGNAME;
     }
 
-    return buffer;
-}
-
-static wchar_t *
-get_process_name()
-{
-    DWORD bufferLen = MAX_PATH;
-    DWORD len = bufferLen;
-    wchar_t *r = NULL;
-
-    while (!r) {
-        r = (wchar_t *)malloc(bufferLen * sizeof(wchar_t));
-        if (!r) {
-            Py_FatalError("out of memory");
-            return NULL;
-        }
-        len = GetModuleFileNameW(NULL, r, bufferLen);
-        if (len == 0) {
-            free((void *)r);
-            return NULL;
-        } else if (len == bufferLen &&
-                   GetLastError() == ERROR_INSUFFICIENT_BUFFER) {
-            free(r);
-            r = NULL;
-            bufferLen *= 2;
+    if (!home.empty()) {
+        PyConfig_SetString(config, &config->home, home.c_str());
+        // XXX: Why doesn't setting prefix and exec_prefix work?
+        //PyConfig_SetString(config, &config->prefix, home.c_str());
+        //PyConfig_SetString(config, &config->exec_prefix, home.c_str());
```

Comment:
@vstinner I expected to be able to just override prefix and exec_prefix, but they don't seem to flow through properly?

Reply:
I don't know: it works for me (see my other comment). config_calculate_pathconfig() should leave config->exec_prefix unmodified if it's already set explicitly.

Later, pycore_init_import_warnings() calls _PyConfig_SetPathConfig() which copies config->exec_prefix into _Py_path_config.exec_prefix.